keystore/v2/keystore/asn1/asn1.go

```diff
@@ -0,0 +1,260 @@
+func UnmarshalVerifiedContainer(data []byte) (*VerifiedContainer, error) {
+	container := new(VerifiedContainer)
+	rest, err := asn1.Unmarshal(data, container)
+	if err != nil {
+		return nil, err
+	}
+	if len(rest) != 0 {
+		return nil, ErrExtraData
+	}
+	return container, nil
+}
+
+// We have to duplicate data definitions for containers because ASN.1's ANY type
+// is represented by interface{} on input and asn1.RawValue on output.
+// We cannot join them in a single structure to avoid duplication, unfortunately.
+
+// SignedPayload contains payload for SignedContainer before marshaling.
+type SignedPayload struct {
+	ContentType  ContentType
+	Version      int
+	LastModified time.Time `asn1:"utc"`
+	Data         interface{}
+}
+
+// Marshal into bytes.
+func (payload *SignedPayload) Marshal() ([]byte, error) {
+	return asn1.Marshal(*payload)
+}
+
+// VerifiedPayload contains payload of VerifiedContainer after unmarshaling.
+type VerifiedPayload struct {
+	RawContent   asn1.RawContent
+	ContentType  ContentType
+	Version      int
+	LastModified time.Time `asn1:"utc"`
+	Data         asn1.RawValue
+}
+
+// ContentType identifies "Data" field of SignedContainer objects.
+type ContentType asn1.Enumerated
+
+// Common ContentType constants:
+const (
+	TypeKeyRing ContentType = iota + 1
+	TypeKeyDirectory
+	TypeDirectKeyDirectory
+)
+
+// Common Version constants:
+const (
+	KeyRingVersion2 = 2
+)
+
+// Signature for SignedContainer. A container can have multiple signatures
+// made with different algorithms, enabling future-proofing, extensibility,
+// and collision resistance. Signatures are computed for the "Payload" of
+// SignedContainer, usually with HMAC keyed by the key store master key.
+// The signing algorithm is indicated by the "Algorithm" field.
+type Signature struct {
+	Algorithm asn1.ObjectIdentifier
+	Signature []byte
+}
+
+// DirectKeyDirectory contains key rings or other key directories.
+// It has a name by which it can be referred to.
+// Child key rings and directories are contained as immediate objects.
+type DirectKeyDirectory struct {
+	Name     LikelyUTF8String
+	KeyRings []KeyRing            `asn1:"set,optional,tag:1"`
+	Children []DirectKeyDirectory `asn1:"set,optional,tag:2"`
+}
+
+// KeyDirectory contains key rings or other key directories.
+// It has a name by which it can be referred to.
+type KeyDirectory struct {
+	Name     LikelyUTF8String
+	KeyRings []KeyRingReference      `asn1:"set,optional,tag:1"`
+	Children []KeyDirectoryReference `asn1:"set,optional,tag:2"`
+}
+
+// UnmarshalKeyDirectory constructs a KeyDirectory from serialized representation.
+func UnmarshalKeyDirectory(data []byte) (*KeyDirectory, error) {
+	container := new(KeyDirectory)
+	rest, err := asn1.Unmarshal(data, container)
+	if err != nil {
+		return nil, err
+	}
+	if len(rest) != 0 {
+		return nil, ErrExtraData
+	}
+	return container, nil
+}
+
+// KeyRingReference to a child key ring, not included into KeyDirectory object directly.
```

Comment:
> can you explain for what and how will be used such references?

It’s a placeholder for currently unimplemented “key directories”. The idea is that each directory will contain a list of `KeyRingReference` and `KeyDirectoryReference` for its direct children. These structures contain names and signatures of objects, so signing a key directory signs its children content too:

```
KeyDirectory {
    name: "ClientAlice"
    rings: [
        KeyRingReference {
            name: "ClientAlice/storage",
            signature: "12345"
        }
    ],
    signature: "67890",
)
```

This should allow us to verify integrity of the key store as a whole, not only of individual key rings. Currently it is possible to remove a keyring entirely, or replace it with any previously signed version without key store noticing it. Key directories will prevent that because they will track what key rings they contain and what data they store.